src/main/java/org/opensearch/knn/index/query/FilterIdsSelector.java

```diff
@@ -0,0 +1,105 @@
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.query;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.BitSetIterator;
+import org.apache.lucene.util.FixedBitSet;
+import org.apache.lucene.util.RamUsageEstimator;
+
+import java.io.IOException;
+
+/**
+ * Util Class for filter ids selector
+ */
+@AllArgsConstructor
+@Getter
+public class FilterIdsSelector {
+
+    /**
+     * When do ann query with filters, there are two types:
+     * BitMap using FixedBitSet, BATCH using a long array stands for filter result docids.
+     */
+    @AllArgsConstructor
+    @Getter
+    public enum FilterIdsSelectorType {
+        BITMAP(0),
+        BATCH(1);
+
+        private final int value;
+    }
+
+    long[] filterIds;
+    FilterIdsSelectorType filterType;
+
+    private static final long SINGLE_ELEMENT_ARRAY_BYTES_USED = RamUsageEstimator.sizeOf(new long[1]);
+
+    /**
+     * This function takes a call on what ID Selector to use:
+     * https://github.com/facebookresearch/faiss/wiki/Setting-search-parameters-for-one-query#idselectorarray-idselectorbatch-and-idselectorbitmap
+     *
+     * class	       storage	lookup     construction(Opensearch + Faiss)
+     * IDSelectorArray	O(k)	O(k)          O(2k)
+     * IDSelectorBatch	O(k)	O(1)          O(2k)
+     * IDSelectorBitmap	O(n/8)	O(1)          O(k) n is the max value of id in the index
+     *
+     * TODO: We need to ideally decide when we can take another hit of K iterations in latency. Some facts:
+     * an OpenSearch Index can have max segment size as 5GB which, which on a vector with dimension of 128 boils down to
+     * 7.5M vectors.
+     * Ref: https://opensearch.org/docs/latest/search-plugins/knn/knn-index/#hnsw-memory-estimation
+     * M = 16
+     * Dimension = 128
+     * (1.1 * ( 4 * 128 + 8 * 16) * 7500000)/(1024*1024*1024) ~ 4.9GB
+     * Ids are sequential in a Segment which means for IDSelectorBitmap total size if the max ID has value of 7.5M will be
+     * 7500000/(8*1024) = 915KBs in worst case. But with larger dimensions this worst case value will decrease.
+     *
+     * With 915KB how many ids can be represented as an array of 64-bit longs : 117,120 ids
+     * So iterating on 117k ids for 1 single pass is also time consuming. So, we are currently concluding to consider only size
+     * as factor. We need to improve on this.
+     *
+     * @param filterIdsBitSet Filter query result docs
+     * @param cardinality The number of bits that are set
+     * @return {@link FilterIdsSelector}
+     */
+    public static FilterIdsSelector getIdSelectorType(BitSet filterIdsBitSet, int cardinality) throws IOException {
```

Comment:
I think we should use builder pattern instead of this static method.
```
  public class FilterIdsSelectorBuilder {
    ...
  }
```

Reply:
@heemin32 I think its fine to use a static method here. Having a separate class will be an overkill here. We can refractor it once we have more than 1 way to build the FilterIdsSelector.